torch/utils/checkpoint.py

```diff
@@ -24,9 +24,22 @@ def check_backward_validity(inputs):
 class CheckpointFunction(torch.autograd.Function):
 
     @staticmethod
-    def forward(ctx, run_function, *args):
+    def forward(ctx, run_function, preserve_rng_state, *args):
         check_backward_validity(args)
         ctx.run_function = run_function
+        ctx.preserve_rng_state = preserve_rng_state
+        if preserve_rng_state:
+            # We can't know if the user will transfer some args from the host
+            # to the device during their run_fn.  Therefore, we stash both
+            # the cpu and cuda rng states unconditionally.
+            #
+            # TODO:
+            # We also can't know if the run_fn will internally move some args to a device
+            # other than the current device, which would require logic to preserve
+            # rng states for those devices as well...but I see no easy way to
+            # handle such cases.
+            ctx.fwd_cpu_rng_state = torch.get_rng_state()
+            ctx.fwd_cuda_rng_state = torch.cuda.get_rng_state()
```

Comment:
this assumes user is using CUDA function. guard this with whether cuda lazy initialization has happened or not... (on current device)

Reply:
Oops, of course, I don't want to trigger initialization of a cuda context if the user doesn't require one.  There remains the (somewhat remote) possibility that the user intends cuda initialization to happen later (within their run_fn).  There's no way for me to anticipate that will happen, though...

Reply:
Hopefully resolved by the latest commit